trunk/source/Library-TraceCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/tracecheckerutils/singletracecheck/AnnotateAndAsserter.java

```diff
@@ -26,114 +28,170 @@
+
+		mSatisfiable = annotateAndAssert(mSSA.getTrace(), partitions);
+		mLogger.info("Assert order " + mAssertCodeBlocksOrder + " issued " + mCheckSat + " check-sat command(s)");
+		mLogger.info("Conjunction of SSA is " + mSatisfiable);
+	}
+
+	private IAssertOrder<L> getAssertOrder(final AssertCodeBlockOrder order) {
+		switch (order.getAssertCodeBlockOrderType()) {
+		case NOT_INCREMENTALLY:
+			return new AssertOrderNotIncrementally<>();
+		case OUTSIDE_LOOP_FIRST1:
+			return new AssertOrderOutsideLoopFirst1<>();
+		case OUTSIDE_LOOP_FIRST2:
+			return new AssertOrderOutsideLoopFirst2<>();
+		case INSIDE_LOOP_FIRST1:
+			return new AssertOrderInsideLoopFirst1<>();
+		case MIX_INSIDE_OUTSIDE:
+			return new AssertOrderMixInsideOutside<>();
+		case TERMS_WITH_SMALL_CONSTANTS_FIRST:
+			return new AssertOrderSmallConstantsFirst<>();
+		case SMT_FEATURE_HEURISTIC:
+			return new AssertOrderSmtFeatureHeuristic<>(order.getSmtFeatureHeuristicScoringMethod(),
+					order.getSmtFeatureHeuristicNumPartitions(), order.getSmtFeatureHeuristicThreshold(),
+					order.getSmtFeatureHeuristicPartitioningType(), mLogger);
+		default:
+			throw new AssertionError("unknown heuristic " + order);
+		}
+	}
+
+	private LBool annotateAndAssert(final NestedWord<? extends IAction> trace, final List<Set<Integer>> partitions) {
+		LBool sat = null;
+		boolean isFirstIteration = true;
+		for (final Set<Integer> partition : partitions) {
+			buildAnnotatedSsaAndAssertTermsWithPriorizedOrder(trace, partition, isFirstIteration);
+			mCheckSat++;
+			sat = mMgdScriptTc.getScript().checkSat();
+			// Report benchmarks
+			mTcbg.reportNewCheckSat();
+			mTcbg.reportNewAssertedCodeBlocks(partition.size());
+			if (sat == LBool.UNSAT) {
+				return sat;
+			}
+			isFirstIteration = false;
+		}
+		return sat;
+	}
+
+	/**
+	 * Annotate and assert every statement <i>i</i> from the given trace, such that <i>i</i> is an element of the given
+	 * integer set stmtsToAssert.
+	 */
+	private void buildAnnotatedSsaAndAssertTermsWithPriorizedOrder(final NestedWord<? extends IAction> trace,
+			final Set<Integer> stmtsToAssert, final boolean assertPendingContexts) {
+		for (final Integer i : stmtsToAssert) {
+			if (trace.isCallPosition(i)) {
 				mAnnotSSA.setGlobalVarAssignmentAtPos(i,
 						mAnnotateAndAssertCodeBlocks.annotateAndAssertGlobalVarAssignemntCall(i));
 				mAnnotSSA.setLocalVarAssignmentAtPos(i,
 						mAnnotateAndAssertCodeBlocks.annotateAndAssertLocalVarAssignemntCall(i));
 				mAnnotSSA.setOldVarAssignmentAtPos(i,
 						mAnnotateAndAssertCodeBlocks.annotateAndAssertOldVarAssignemntCall(i));
 			} else {
-				if (mTrace.isReturnPosition(i) && mTrace.isPendingReturn(i)) {
-					pendingReturnPositions.add(i);
-				}
 				mAnnotSSA.setFormulaAtNonCallPos(i, mAnnotateAndAssertCodeBlocks.annotateAndAssertNonCall(i));
 			}
 		}
 
-		assert callPositions.containsAll(mTrace.getCallPositions());
-		assert mTrace.getCallPositions().containsAll(callPositions);
```

Comment:
Did you intend to remove these checks?

Reply:
Oh, that's interesting: These checks were not present in `AnnotateAndAsserterWithStmtOrderPrioritization` (which is the basis for the new `AnnotateAndAsserter`). Therefore I also removed the variable `callPositions`, because it was unused. But I think it does not hurt to add these checks again.

Reply:
Wait, these assertions state that all calls of the trace are already asserted. This is only reasonable, if the whole SSA is asserted here. We could keep the second assertion, but it would be trivially satisfied, since for all `i`, if `trace.isCallPosition(i)` is true, then `i` is element of `trace.getCallPositions()`.